src/fcmp_pp/fcmp_pp_types.h

```diff
@@ -67,20 +68,78 @@ struct HeliosT final
 using OutputBytes = ::OutputBytes;
 using OutputChunk = ::OutputSlice;
 //----------------------------------------------------------------------------------------------------------------------
+// Define C++ type here so it can be used in FFI types
+using FcmpPpProof = std::vector<uint8_t>;
+//----------------------------------------------------------------------------------------------------------------------
+//----------------------------------------------------------------------------------------------------------------------
+// FFI types
+//----------------------------------------------------------------------------------------------------------------------
+// FFI types instantiated on the Rust side must be destroyed back on the Rust side. We wrap them in a unique ptr with a
+// custom deleter that calls the respective Rust destroy fn.
+#define DEFINE_FCMP_FFI_TYPE(raw_t, cpp_fn)                                      \
+    struct raw_t##Deleter { void operator()(raw_t##Unsafe *p) const noexcept; }; \
+    using raw_t = std::unique_ptr<raw_t##Unsafe, raw_t##Deleter>;                \
+    raw_t cpp_fn;
+
+// Sometimes a shared ptr is necessary so that we can reference the same copy of underlying data in multiple places.
+#define DEFINE_FCMP_FFI_SHARED_TYPE(raw_t, cpp_fn1, cpp_fn2) \
+    using raw_t = std::shared_ptr<raw_t##Unsafe>;            \
+    raw_t cpp_fn1;                                           \
+    raw_t cpp_fn2;
```

Comment:
I personally don't see the point in putting these functions in the macro. It makes the macros harder to read, it makes the function definitions less clear, and it really isn't any less writing.